### PR TITLE
Json Schema WIP

### PR DESCRIPTION
**Pull request by gecage952**
PR for redoing Shaman code using DataElements annotations.

### CODE REVIEW

**Comment by gecage952** [comment]
I'm sure there a lot of things that can be improved and a lot of style and formatting improvements that should be made, so it would be helpful to get all that feedback if anyone has the time.

File: org.eclipse.ice.dev/org.eclipse.ice.dev.jsonschemaconverter/src/main/java/org/eclipse/ice/dev/jsonschemaconverter/JsonSchemaConverter.java [hunk 1]
```diff
@@ -0,0 +1,427 @@
...
+	   }
+   }
+
+   /**
+    * Gives the string representation of the type of input string.
+    * Used when input string may represent an int, float, boolean, etc.
+    * @param input string that may represent a type other than string
+    * @return the type of the input as a string
+    */
+   public static String getTypeAsString(String input) {
```
Comment (dbluhm) [inline org.eclipse.ice.dev/org.eclipse.ice.dev.jsonschemaconverter/src/main/java/org/eclipse/ice/dev/jsonschemaconverter/JsonSchemaConverter.java:373]: We might want to move these types of functions to some sort of Java Source formatter at some point. I had to do similar things with the `DefaultFields` class and having some of these available to templates or annotation processors may come in handy. I think here is good for now -- just taking note of it.

File: org.eclipse.ice.dev/org.eclipse.ice.dev.jsonschemaconverter/src/test/java/org/eclipse/ice/dev/jsonschemaconverter/JsonSchemaConverterTest.java [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+	
+	@AfterAll
+	public static void teardown() throws IOException {
+		Files.delete(destination.getParent().resolve(JSON_RESULT));
+		Files.delete(destination.getParent().resolve(GENERATED_INTERFACE));
+		Files.delete(destination.getParent().resolve(GENERATED_IMPLEMENTATION));
+		Files.delete(destination.getParent().resolve(GENERATED_PROP_INTERFACE));
+		Files.delete(destination.getParent().resolve(GENERATED_PROP_IMPLEMENTATION));
+	}
+
```
Comment (dbluhm) [inline org.eclipse.ice.dev/org.eclipse.ice.dev.jsonschemaconverter/src/test/java/org/eclipse/ice/dev/jsonschemaconverter/JsonSchemaConverterTest.java:62]: Going along with my comments on `JsonSchemaConverter` some tests with nasty file names might be good.

File: org.eclipse.ice.dev/org.eclipse.ice.dev.pojofromjson/src/main/java/org/eclipse/ice/dev/pojofromjson/PojoFromJson.java [hunk 3]
```diff
@@ -76,38 +76,47 @@ public static void handleInputJson(
...
+					)) {
+						ImplementationWriter.builder()
+							.packageName(pojo.getPackageName())
+							.interfaceName(pojo.getElement())
+							.className(pojo.getImplementation())
+							.fields(fields)
+							.types(fields.getTypes())
+							.build()
+							.write(elementImpl);
+					}
```
Comment (dbluhm) [inline org.eclipse.ice.dev/org.eclipse.ice.dev.pojofromjson/src/main/java/org/eclipse/ice/dev/pojofromjson/PojoFromJson.java:119]: Seems to be some strange formatting stuff going on here.

**Review by dbluhm — COMMENTED** [review]
Nice work! I made a few comments and suggestions. If you ignore my `.map()` suggestions, I won't be offended lol but using the file name as the package name might have a few potential issues that I noted.

File: org.eclipse.ice.data.neutrons/src/main/java/org/eclipse/ice/data/neutrons/RunSpec.java [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+	/**
+	 * user definable label
+	 */
+	@DataField private String emptyLabel;
+
+	/**
+	 * a hacky place to store just a value, not associated with any run, typically
+	 * setting the transmission to _e.g._ 0.9
+	 */
+	@DataField	String value;
```
[the remark this answers is not in the document]
Comment (gecage952) [reply org.eclipse.ice.data.neutrons/src/main/java/org/eclipse/ice/data/neutrons/RunSpec.java:123]: I think the reasoning is that I erased "private" inadvertently.

File: org.eclipse.ice.dev/org.eclipse.ice.dev.jsonschemaconverter/src/main/java/org/eclipse/ice/dev/jsonschemaconverter/JsonSchemaConverter.java [hunk 5]
```diff
@@ -0,0 +1,427 @@
...
+    * @param filePath string representing path to input file
+    * @throws JsonParseException On failure to parse the input JSON schema file
+    * @throws JsonMappingException On failure to map the JSON schema to Map<String, Object>
+    * @throws IOException On failure to write files
+    */
+   public static void handleInputJson(InputStream is, Path filePath) throws IOException {
+	   	Map<String, Object> map = mapper.readValue(is, new TypeReference<Map<String,Object>>(){});
+	   	List<PojoOutline> jsonArrayOut = new ArrayList<>();
+	   	String fileName = filePath.getFileName().toString();
+	   	fileName = fileName.substring(0, fileName.length() - 5); //remove .json
```
[the remark this answers is not in the document]
Comment (gecage952) [reply org.eclipse.ice.dev/org.eclipse.ice.dev.jsonschemaconverter/src/main/java/org/eclipse/ice/dev/jsonschemaconverter/JsonSchemaConverter.java:139]: I think maybe something like https://docs.oracle.com/javase/8/docs/api/javax/lang/model/SourceVersion.html#isName-java.lang.CharSequence- might work.

File: org.eclipse.ice.dev/org.eclipse.ice.dev.pojofromjson/src/main/java/org/eclipse/ice/dev/pojofromjson/PojoFromJson.java (hunk 3 above)
Comment (gecage952) [reply org.eclipse.ice.dev/org.eclipse.ice.dev.pojofromjson/src/main/java/org/eclipse/ice/dev/pojofromjson/PojoFromJson.java:119]: Yeah, sometimes weird things happen with formatting when I'm moving pieces of code.

**Comment by jayjaybillings** [comment]
What are we waiting on for this PR? Me to review? Also, can someone fix the merge conflicts?

**Comment by gecage952** [comment]
Yeah, I addressed Daniel's comments, so I was just waiting for a review of those latest changes I believe. Also, I can fix the conflicts.

**Comment by jayjaybillings** [comment]
@gecage952 Thanks! @dbluhm You good with this? If so, I'll give it a look.

File: org.eclipse.ice.data.neutrons/frontend/src/org/eclipse/ice/README [hunk 6]
```diff
@@ -0,0 +1 @@
+Place your Vaadin Designer or hand written templates in this folder.
```
Comment (jayjaybillings) [inline org.eclipse.ice.data.neutrons/frontend/src/org/eclipse/ice/README:1]: Is there better text for the README?

File: org.eclipse.ice.data.neutrons/frontend/styles/org/eclipse/ice/README [hunk 7]
```diff
@@ -0,0 +1,3 @@
+Place your optional styles in this folder.
```
Comment (jayjaybillings) [inline org.eclipse.ice.data.neutrons/frontend/styles/org/eclipse/ice/README:1]: Is there better text for this README too?

It is OK to have the default text, but we should consider adding some context for why these are here if we are going to keep them. Something like "This file is auto-generated by our project generators. It is for Vaadin. Blah blah blah..."

File: org.eclipse.ice.data.neutrons/src/test/java/org/eclipse/ice/data/neutrons/AppTest.java [hunk 8]
```diff
@@ -0,0 +1,14 @@
...
+import static org.junit.Assert.fail;
+import org.junit.Test;
+
+public class AppTest {
+   
+	@Test
+    public void shouldAnswerWithTrue() {
+		ReductionConfiguration ex = new ReductionConfigurationImplementation();
+		assertTrue(ex.getNumQBins() == 100);   	
+    }
```
Comment (jayjaybillings) [inline org.eclipse.ice.data.neutrons/src/test/java/org/eclipse/ice/data/neutrons/AppTest.java:13]: Is this a thorough test?

File: org.eclipse.ice.dev/org.eclipse.ice.dev.jsonschemaconverter/src/main/java/org/eclipse/ice/dev/jsonschemaconverter/JsonSchemaConverter.java [hunk 9]
```diff
@@ -0,0 +1,422 @@
...
+   			if (jsonFiles.isEmpty()) {
+   				handleInputJson(System.in, Path.of(output));
+   			}
+   			for (String filePath : jsonFiles) {
+
+   				try (FileInputStream inputJson = 
+   											new FileInputStream(filePath)) {
+   					handleInputJson(inputJson, Path.of(filePath));
+   				}
+   			}
```
Comment (jayjaybillings) [inline org.eclipse.ice.dev/org.eclipse.ice.dev.jsonschemaconverter/src/main/java/org/eclipse/ice/dev/jsonschemaconverter/JsonSchemaConverter.java:122]: Some comments and cleaner formatting would be good here.

File: org.eclipse.ice.dev/org.eclipse.ice.dev.jsonschemaconverter/src/main/java/org/eclipse/ice/dev/jsonschemaconverter/JsonSchemaConverter.java [hunk 10]
```diff
@@ -0,0 +1,422 @@
...
+
+/**
+ * Utility to convert JSON schemas to the JSON format that the
+ * PojoFromJson project accepts
+ * Pieces of code were derived from Daniel Bluhm's
+ * org.eclipse.ice.dev.PojoFromJson.
+ * @author gzi
+ *
+ */
+public class JsonSchemaConverter {
```
Comment (jayjaybillings) [inline org.eclipse.ice.dev/org.eclipse.ice.dev.jsonschemaconverter/src/main/java/org/eclipse/ice/dev/jsonschemaconverter/JsonSchemaConverter.java:38]: Can you run the Eclipse formatter on this please?

File: org.eclipse.ice.dev/org.eclipse.ice.dev.jsonschemaconverter/src/main/java/org/eclipse/ice/dev/jsonschemaconverter/JsonSchemaConverter.java [hunk 11]
```diff
@@ -0,0 +1,427 @@
...
+    * @return boolean representing whether or not input is an int
+    */
+   public static boolean isInt(String input) {
+	   try {
+		   Integer.parseInt(input);
+		   return true;
+	   } catch (Exception e) {
+		   return false;
+	   }
+   }
```
[the remark this answers is not in the document]
Comment (jayjaybillings) [reply org.eclipse.ice.dev/org.eclipse.ice.dev.jsonschemaconverter/src/main/java/org/eclipse/ice/dev/jsonschemaconverter/JsonSchemaConverter.java:340]: Looks like this is fixed...

File: org.eclipse.ice.dev/org.eclipse.ice.dev.jsonschemaconverter/src/main/java/org/eclipse/ice/dev/jsonschemaconverter/JsonSchemaConverter.java (hunk 1 above)
Comment (jayjaybillings) [reply org.eclipse.ice.dev/org.eclipse.ice.dev.jsonschemaconverter/src/main/java/org/eclipse/ice/dev/jsonschemaconverter/JsonSchemaConverter.java:348]: Agreed.

File: org.eclipse.ice.dev/org.eclipse.ice.dev.jsonschemaconverter/src/test/java/org/eclipse/ice/dev/jsonschemaconverter/JsonSchemaConverterTest.java (hunk 2 above)
Comment (jayjaybillings) [reply org.eclipse.ice.dev/org.eclipse.ice.dev.jsonschemaconverter/src/test/java/org/eclipse/ice/dev/jsonschemaconverter/JsonSchemaConverterTest.java:62]: Agreed.

**Review by jayjaybillings — CHANGES_REQUESTED** [review]
Minor changes requested. Good work.

**Comment by dbluhm** [comment]
I think the tests for nasty filenames is the last comment I had. Marked everything else as resolved :slightly_smiling_face:

**Comment by jayjaybillings** [comment]
@gecage952 Did testing for nasty file names make it in there?

**Comment by gecage952** [comment]
I actually didn't address that yet. I'll add more testing around that while I'm fixing the merge conflicts.

**Review by jayjaybillings — APPROVED** [review]

**Comment by jayjaybillings** [comment]
Approved and accepted. Thanks very much for your contribution! 

@gecage952 is an employee of a member company, UT-Battelle LLC and all other checks pass.

**Comment by gecage952** [comment]
I actually haven't pushed the testing changes quite yet. Those commits were just fixing the merge conflicts.

**Comment by dbluhm** [comment]
Ah, woops! @gecage952, please feel free to open another PR with those changes when you're ready

**Pull request merged**